server/plugin.go

```diff
@@ -18,11 +35,237 @@ type Plugin struct {
 	// configuration is the active plugin configuration. Consult getConfiguration and
 	// setConfiguration for usage.
 	configuration *configuration
+
+	msteamsAppClientMutex sync.Mutex
+	msteamsAppClient      msteams.Client
+	msteamsBotClientMutex sync.Mutex
+	msteamsBotClient      msteams.Client
+
+	botID  string
+	userID string
+
+	subscriptionsToLinksMutex sync.Mutex
+	subscriptionsToLinks      map[string]ChannelLink
+	channelsLinked            map[string]ChannelLink
+
+	stopSubscriptions func()
 }
 
-// ServeHTTP demonstrates a plugin that handles HTTP requests by greeting the world.
 func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
-	fmt.Fprint(w, "Hello, world!")
+	api := NewAPI(p)
+	api.ServeHTTP(w, r)
+}
+
+func (p *Plugin) getURL() string {
+	config := p.API.GetConfig()
+	if strings.HasSuffix(*config.ServiceSettings.SiteURL, "/") {
+		return *config.ServiceSettings.SiteURL + "plugins/" + pluginID
+	}
+	return *config.ServiceSettings.SiteURL + "/plugins/" + pluginID
+}
+
+func (p *Plugin) connectTeamsAppClient() error {
+	p.msteamsAppClientMutex.Lock()
+	if p.msteamsAppClient == nil {
+		p.msteamsAppClient = msteams.NewApp(
+			p.configuration.TenantId,
+			p.configuration.ClientId,
+			p.configuration.ClientSecret,
+		)
+	}
+	err := p.msteamsAppClient.Connect()
+	if err != nil {
+		p.API.LogError("Unable to connect to the app client", "error", err)
+		return err
+	}
+	p.msteamsAppClientMutex.Unlock()
+	return nil
 }
 
-// See https://developers.mattermost.com/extend/plugins/server/reference/
+func (p *Plugin) connectTeamsBotClient() error {
+	p.msteamsBotClientMutex.Lock()
+	if p.msteamsBotClient == nil {
+		p.msteamsBotClient = msteams.NewBot(
+			p.configuration.TenantId,
+			p.configuration.ClientId,
+			p.configuration.ClientSecret,
+			p.configuration.BotUsername,
+			p.configuration.BotPassword,
+		)
+	}
+	err := p.msteamsBotClient.Connect()
+	if err != nil {
+		p.API.LogError("Unable to connect to the bot client", "error", err)
+		return err
+	}
+	p.msteamsBotClientMutex.Unlock()
```

Comment:
Should we instead `defer` this call above, to avoid the error returning in the previous block?

Reply:
Totaly right, it wasn't like that before, I'll add the defer there.